helics/capi.py

```diff
@@ -717,9 +718,33 @@ class HelicsFederateState(IntEnum):
 helics_state_pending_finalize = HelicsFederateState.PENDING_FINALIZE
 
 
+def generate_cleanup_callback(obj):
+
+    if issubclass(obj.__class__, HelicsFederate):
+        f = loadSym("helicsFederateFree")
+    elif issubclass(obj.__class__, HelicsFederateInfo):
+        f = loadSym("helicsFederateInfoFree")
+    elif issubclass(obj.__class__, HelicsBroker):
+        f = loadSym("helicsBrokerFree")
+    elif issubclass(obj.__class__, HelicsCore):
+        f = loadSym("helicsCoreFree")
+    elif issubclass(obj.__class__, HelicsQuery):
+        f = loadSym("helicsQueryFree")
+    # elif issubclass(obj.__class__, HelicsMessage):
+    #     f = loadSym("helicsMessageFree")
+    else:
+        f = lambda handle: None
+
+    def cleanup(handle):
+        f(handle)
```

Comment:
Any thoughts on calling `helicsCleanupLibrary()` here? The call to `helicsCleanupLibrary` is what took care of the state corruption in my tests. Seeing as this will only be called when objects are destroyed, it *shouldn't* be called too frequently to create performance problems, right?

Reply:
If we had code like this, then wouldn't `helicsCleanupLibrary` be called when `utils()` exists scope?

```
def utils():
  q = h.helicsCreateQuery()

def main():
  utils()
  ...
```

Reply:
Correct, and that seems useful to me. The docs say:

>     Function to do some housekeeping work.
>
>     This runs some cleanup routines and tries to close out any residual thread that haven’t been shutdown yet.

I would presume this "housekeeping" is checking for orphaned threads and the like and shouldn't close out active threads? If this is not the case and a call to `helicsCleanupLibrary` will disrupt active threads, then I 100% agree with you - it would be a bad idea to add this cleanup call to the finalizer.

Reply:
@phlptp Thoughts? The tests all pass when I add it in there.

Reply:
I've pushed a commit https://github.com/GMLC-TDC/pyhelics/pull/26/commits/3f32261177b63d8e933f9d4a84ed36ad49609496 with this change to see what happens on github actions.

Reply:
Is the state corruption in your tests reproducible? Would you be able to create an issue for it on the HELICS repo?

Reply:
It was consistent on my machine, yes, but I haven't asked any colleagues to try it locally.

I just got the go-ahead today for us (Utilidata developers) to contribute to pyhelics. I'll make some time tomorrow to reproduce the problem and submit an issue.

We'll also plan on forking pyhelics soon and submitting PRs. Perhaps we should schedule a meeting to discuss how we can make this interaction as smooth as possible.

Reply:
I'll go ahead and merge this now. If you are able reproducible your state corruption, if you can file an issue on the main repo that'll be great!

Reply:
Ah, sorry, I filed it on this repo before I saw this comment! Let me know if you'd like me to relocate it.

Reply:
Also, I suspect the tests I gave in that issue will pass off-the-bat in a `pyhelics` version with these updates merged in.